.github/scripts/tests.sh

```diff
@@ -0,0 +1,280 @@
+#!/bin/bash
+
+source functions.sh
+
+function check_right_cases() {
+    declare -a right_cases=("${!2}")
+    local ret=0
+
+    for el in "${right_cases[@]}"; do
+        echo $el | $1 /dev/stdin > /dev/null
+        if [[ $? != 0 ]]; then
+            echo "❌ Right cases check not passed: $el"
+            ret=1
+        fi
+    done
+
+    return $ret
+}
+
+function check_wrong_cases() {
+    declare -a wrong_cases=("${!2}")
+    local ret=0
+
+    for el in "${wrong_cases[@]}"; do
+        echo $el | $1 /dev/stdin > /dev/null
+        if [[ $? -eq 0 ]]; then
+            echo "❌ Wrong cases check not passed: $el"
+            ret=1
+        fi
+    done
+
+    return $ret
+}
+
+function run_test() {
+    local right_cases=("${!3}")
+    local wrong_cases=("${!4}")
+    local ret=0
+
+    echo "Testing $1 🔎"
+
+    check_right_cases $2 right_cases[@]
+    if [[ $? != 0 ]]; then
+        ret=1
+    fi
+
+    check_wrong_cases $2 wrong_cases[@]
+    if [[ $? != 0 ]]; then
+        ret=1
+    fi
+
+    if [[ $ret != 0 ]]; then
+        echo "❌ $1 check not passed"
+    else
+        echo "✅ $1 check passed"
+    fi
+
+    echo
+    return $ret
+}
+
+function test_shared_ptr_copy_constructor() {
+    local right_copy_ctors=("shared_ptr(const shared_ptr& other)"
+                            "shared_ptr(shared_ptr const& ptr)"
+                            "shared_ptr(const shared_ptr&)"
+                            "shared_ptr(shared_ptr const&)"
+                            "shared_ptr(
+                                const shared_ptr& other)"
+                            "shared_ptr(shared_ptr const &sth)"
+                            "shared_ptr(shared_ptr const & sth)"
+                            "shared_ptr(const shared_ptr &sth)"
+                            "shared_ptr(const shared_ptr & sth)")
+
+    local wrong_copy_ctors=("shared_ptr(const shared_ptr&& other)"
+                            "shared_ptr(shared_ptr&&)"
+                            "shared(const shared_ptr& other)"
+                            "shared_ptr(shared_ptr& ptr)")
+
+    run_test "copy ctor" check_shared_ptr_copy_constructor right_copy_ctors[@] wrong_copy_ctors[@]
+    return $?
```

Comment:
can we merge those 2 lines?

```suggestion
    return run_test "copy ctor" check_shared_ptr_copy_constructor right_copy_ctors[@] wrong_copy_ctors[@]
```

or the behavior is different then?

Reply:
As far as I know, `return` in bash can only return exit codes, which can only be accessed using `$?`. Suggested version won't work at all.